### PR TITLE
Feature: Payment Information Banner

### PR DESCRIPTION
**Pull request by JoshuaHungDinh**
## Description

This PR includes the functionality to manage, update and display the Payment information banner for the donation detail page.

Since the same Window object is being used for the ListTable and the AdminPage adjustments to the DonationsListTable were necessary to put this together. The Window declaration for "GiveDonations" has been moved outside of the ListTable component into a common space for both the Table and the Admin page. The types have also been adjusted throughout the ListTable components to reflect the conditional options being provided per display .

The payment banner is made up of a Legend component that handles the display of the title + the status field and 5 other editable fields that are all maintained with react-hook-form.  Users are now able to select a donation from the table and update the details associated with that donation. The data populated into the payment banner reflects the donations current live data.

## Fields 

- Amount: manages the current donations amount with a modal that allows you to set the amount.
- Fee Recovered: manages the current donations fee amount with a modal that allows you to set the amount.
- Forms: manages a list of donation forms that the donation is assigned to with a searchable dropdown.
- DatePicker: manages the selected date that the donation was made with a calendar component from react-dates.
- TimePicker: manages the time of the donation with regular inputs.
- DonationMethod: A static UI display of the donation method based on gateway.
- Status: manages the donation status with a provided list of options to a react-select dropdown component.

## Affects

Donations ListTable

Donation Admin Page

## Visuals

https://user-images.githubusercontent.com/75056371/229931143-54881592-65db-4a12-9fd0-fae63ac1d25e.mov

## Testing Instructions

- Select any donation from the Donation table.
- Edit the various donation fields and save the changes.
- Refresh the page or head back to the Donation table to verify the changes were saved.

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [ ] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

### CODE REVIEW

**Comment by JasonTheAdams** [comment]
@JoshuaHungDinh This is failing static analysis. Please fix those and then request a review.

File: src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/HiddenFields/index.tsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+export default function HiddenFields() {
+    const {register} = useFormContext();
+    return (
+        <>
+            <input hidden {...register('amount')} />
+            <input hidden {...register('feeAmountRecovered')} />
+            <input hidden {...register('createdAt')} />
+        </>
+    );
+}
```
Comment (JasonTheAdams) [inline src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/HiddenFields/index.tsx:16]: I would just move these hidden fields to the container components so they're coupled together with their corresponding parts.

File: src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/TimePicker/index.tsx [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+type TimeFieldProps = {
+    isOpen: boolean;
+    closeFields: (newDate: Date | null) => void;
+    initialTime: Date;
+};
+
+export function TimeFields({isOpen, closeFields, initialTime}: TimeFieldProps) {
+    const [hours, setHours] = useState<number>(Number(format(initialTime, initialTime.getHours() >= 12 ? 'h' : 'h')));
+    const [minutes, setMinutes] = useState<number>(Number(format(initialTime, String(initialTime.getMinutes()))));
+    const [ampm, setAmpm] = useState<string>(format(initialTime, 'h:mm a').split(' ')[1]);
```
Comment (JasonTheAdams) [inline src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/TimePicker/index.tsx:57]: Fun fact, the am/pm unit is called the meridiem. 🤓

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
Not quite done, but some feedback so far!

File: src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/TimePicker/index.tsx (hunk 2 above)
Comment (JoshuaHungDinh) [reply src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/TimePicker/index.tsx]: Fancy..  I like it! Let's update this.  https://github.com/impress-org/givewp/pull/6752/commits/3bf123272f798e5c17af62674640d3186137bb0e

File: src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/HiddenFields/index.tsx (hunk 1 above)
Comment (JoshuaHungDinh) [reply src/Donations/resources/DonationDetails/app/components/PaymentInformation/Fields/HiddenFields/index.tsx]: Sounds good. I chose to place createdAt with the datepicker even though its split between that and the timepicker. Resolved: https://github.com/impress-org/givewp/pull/6752/commits/380558d201fc7ed7e7385c49d1ce15b93e1daf30

File: src/Donations/resources/DonationDetails/app/components/FormTemplate/index.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+ * @unreleased
+ */
+
+export default function FormTemplate() {
+    return (
+        <>
+            <PaymentInformation />
+        </>
+    );
+}
```
[the remark this answers is not in the document]
Comment (JoshuaHungDinh) [reply src/Donations/resources/DonationDetails/app/components/FormTemplate/index.tsx:16]: I have it eventually representing the layout in my following branches.  
export default function FormTemplate() {
    return (
        <React.Fragment>
            <PaymentInformation />

            <Container>
                <LeftContainer>
                    <DonorDetails />
                    <BillingAddress />
                </LeftContainer>

                <RightContainer>
                    <Subscription />
                    <DonorComments />
                    <DonorNotes />{' '}
                </RightContainer>
            </Container>
        </React.Fragment>
    );
}

File: src/Donations/resources/DonationDetails/app/css/style.scss [hunk 4]
```diff
@@ -1,6 +1,11 @@
 //TODO: Load variables with give-foundation
 :root {
     //Color Palette
+    --gutenberg-blue: #007cba;
+    --wp-blue-blue-50: #2271b1;
+    --wp-blue-blue-40: #3582c4;
+    --wp-blue-blue-0: #f0f6fc;
+    --wp-blue-blue-60: #135e96;
```
[the remark this answers is not in the document]
Comment (JoshuaHungDinh) [reply src/Donations/resources/DonationDetails/app/css/style.scss]: Yes! This is a recent change the blue reflects tones of the official WordPress blue. We should probably come up with a better name. I was planning on including it into the foundation. This css file was our temporary solution. We need to swap this out with the official Give styling package.

https://github.com/impress-org/givewp/pull/6752/commits/e301a63cb6fcb8194d3d9a12a63cae5644a43bc2

File: src/Views/Components/AdminUI/Button/index.tsx [hunk 5]
```diff
@@ -24,7 +24,7 @@ const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
     ({children, type = 'button', classname, variant = 'primary', size = 'normal', disabled, ...props}, ref) => (
         <button
             ref={ref}
-            disabled={disabled}
+            disabled={typeof disabled === 'undefined' || disabled}
```
[the remark this answers is not in the document]
Comment (JoshuaHungDinh) [reply src/Views/Components/AdminUI/Button/index.tsx]: Resolved: https://github.com/impress-org/givewp/pull/6752/commits/f069d7b1476d22ec2cfd76137c7005eebfa2aea3

File: src/Donations/resources/DonationDetails/app/css/style.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (JasonTheAdams) [reply src/Donations/resources/DonationDetails/app/css/style.scss]: Got it. Let's talk about this more later. Please bring this up at the next Monday dev call.

File: src/Donations/resources/DonationDetails/app/utilities/formatter.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    }).format(amount);
+}
+
+/**
+ * @unreleased
+ */
+
+export function parseAmountValue(num: number): number {
+    return num ? formatDecimalPlacement(num) : 0;
+}
```
[the remark this answers is not in the document]
Comment (JoshuaHungDinh) [reply src/Donations/resources/DonationDetails/app/utilities/formatter.ts]: Resolved: https://github.com/impress-org/givewp/pull/6752/commits/5a7c1e5ead597265ba13b0e9c9d0566fcf52aa74

File: src/Donations/DonationsAdminPage.php [hunk 7]
```diff
@@ -78,6 +78,7 @@ public function loadScripts()
                 $data,
                 [
                     'apiRoot' => esc_url_raw(rest_url('give-api/v2/admin/donation')),
+                    'currencyFormat' => give_get_currency_formatting_settings(),
```
[the remark this answers is not in the document]
Comment (JoshuaHungDinh) [reply src/Donations/DonationsAdminPage.php:81]: Resolved: https://github.com/impress-org/givewp/pull/6752/commits/5a7c1e5ead597265ba13b0e9c9d0566fcf52aa74

File: src/Donations/resources/DonationDetails/app/css/style.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoshuaHungDinh) [reply src/Donations/resources/DonationDetails/app/css/style.scss]: Yea that sounds like a good idea. I'll do that.

**Comment by JoshuaHungDinh** [comment]
Hey @JasonTheAdams, I have addressed your feedback. Can you please review this again?

File: src/Views/Components/AdminUI/SearchSelector/index.tsx [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+import styles from './style.module.scss';
+import {StyleConfig} from './StyleConfig';
+
+/**
+ *
+ * @unreleased
+ */
+
+export type SearchSelector = {
```
Comment (JasonTheAdams) [inline src/Views/Components/AdminUI/SearchSelector/index.tsx:16]:
```suggestion
export type SearchSelectorProps = {
```

File: src/Views/Components/AdminUI/SearchSelector/index.tsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        label: string;
+    }>;
+    defaultLabel?: string;
+    name: string;
+    placeholder: string;
+};
+
+export default function SearchSelector({options, name, placeholder}: SearchSelector) {
+    const {control} = useFormContext();
+    const [focus, setFocus] = useState(false);
```
Comment (JasonTheAdams) [inline src/Views/Components/AdminUI/SearchSelector/index.tsx:28]:
```suggestion
    const [focus, setFocus] = useState<boolean>(false);
```

File: src/Views/Components/AdminUI/Toast/index.tsx [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+export type ToastProps = {
+    resultType: 'success' | 'error' | null;
+    resultMessage: string;
+    closeMessage: () => void;
+    showMessage: boolean;
+};
+
+export default function Toast({resultType, resultMessage, showMessage, closeMessage}: ToastProps) {
+    const [isActive, setIsActive] = useState(false);
```
Comment (JasonTheAdams) [inline src/Views/Components/AdminUI/Toast/index.tsx:16]:
```suggestion
    const [isActive, setIsActive] = useState<boolean>(false);
```

File: src/Views/Components/AdminUI/api.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import {useState} from 'react';
+
+const PostRequest = (endpoint: string, apiNonce: string, successMessage: string, errorMessage: string) => {
+    const [result, setResult] = useState({type: null, message: ''});
```
Comment (JasonTheAdams) [inline src/Views/Components/AdminUI/api.ts:4]: Is `PostRequest` used as a component, a hook, or a function? We're using a hook here, so it must be one of the first two. I would think a hook, in which case the name should be `usePostRequest`.

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
Great updates, @JoshuaHungDinh! Just a few minor suggestions!

File: src/Views/Components/AdminUI/api.ts (hunk 11 above)
Comment (JoshuaHungDinh) [reply src/Views/Components/AdminUI/api.ts:4]: Resolved: https://github.com/impress-org/givewp/pull/6752/commits/9b6b50ee131b40b30fd730d50c2f5debff76df09

File: src/Views/Components/AdminUI/SearchSelector/index.tsx (hunk 8 above)
Comment (JoshuaHungDinh) [reply src/Views/Components/AdminUI/SearchSelector/index.tsx]: Resolved: https://github.com/impress-org/givewp/pull/6752/commits/0d02295bf42cd6049200ed9459d4b969f3c945e6

File: src/Views/Components/AdminUI/SearchSelector/index.tsx (hunk 9 above)
Comment (JoshuaHungDinh) [reply src/Views/Components/AdminUI/SearchSelector/index.tsx]: Resolved: https://github.com/impress-org/givewp/pull/6752/commits/0d02295bf42cd6049200ed9459d4b969f3c945e6

File: src/Views/Components/AdminUI/FormPage/index.tsx [hunk 12]
```diff
@@ -1,64 +1,90 @@
...
-    pageDetails,
-    navigationalOptions,
     children,
     actionConfig,
+    apiNonce,
+    successMessage,
+    errorMessage,
 }: FormPageProps) {
+    const {postData, result} = usePostRequest(endpoint, apiNonce, successMessage, errorMessage);
+    const [showApiMessage, setApiShowMessage] = useState(false);
```
Comment (JasonTheAdams) [inline src/Views/Components/AdminUI/FormPage/index.tsx:49]:
```suggestion
    const [showApiMessage, setApiShowMessage] = useState<boolean>(false);
```

**Review by JasonTheAdams — APPROVED** [review]
Excellent work, @JoshuaHungDinh! I left one super minor suggestion which you can just accept to merge. Otherwise, let's merge this!

**Pull request merged**